### ISSUE TITLE
Testing sutil app

### ISSUE THREAD

**Issue by marcingolenia** [davedawkins/Sutil#48]
Hi there! 
Thank You for great on work on Sutil. I wanted to develop small app using beta version but I struggle to setup tests :( I am hoping to get some help.

1. I've stolen TestFramework.fs from Sutil repository.
2. The test looks like this;
```fsharp
module Tests.Tests

open TestFramework
open Sutil

let tests = testList "Sutil.DOM" [

    // Simplest case
    testCase "Hello World" <| fun () ->
        let app =
            Html.div "Hello World"

        mountTestApp app

        Expect.queryText "div" "Hello World"
        Expect.areEqual(1, 1)
]
```
3. Commands I use (stolen from sutil repo):
```
    "tests-console": "mocha dist/tests -r esm",
    "watch:tests": "dotnet fable watch tests -o dist/tests --runWatch mocha dist/tests -r esm",
    "tests": "dotnet fable tests --run webpack --config webpack.config.tests.js",
    "start:tests": "dotnet fable watch tests --run webpack serve --config webpack.config.tests.js "
```
The best what I get after running `start:tests` is attached on the screen:
![image](https://user-images.githubusercontent.com/5363851/136210491-6215487c-ed5d-4369-88f0-6735ab67b7ad.png)
So it is empty. When I run tests:watch I get:
```bash
> watch:tests
> dotnet fable watch tests -o dist/tests --runWatch mocha dist/tests -r esm

Fable: F# to JS compiler 3.2.9
Thanks to the contributor! @JacobChang
tests> dotnet restore Tests.fsproj
  Determining projects to restore...
  All projects are up-to-date for restore.
Parsing tests/Tests.fsproj...
Initializing F# compiler...
Compiling tests/Tests.fsproj...
F# compilation finished in 5621ms
Fable compilation finished in 1392ms
.> node_modules/.bin/mocha dist/tests -r esm
Watching .
logging:init defaults


  0 passing (1ms)
```
When I run test-console I get;
```bash
> tests-console
> mocha dist/tests -r esm

logging:init defaults


  0 passing (0ms)
```
my webpack.config.tests.js looks like this;
```
var path = require("path");

module.exports = {
    mode: "development",
    entry: "./tests/Tests.fs.js",
    output: {
        path: path.join(__dirname, "./tests/public"),
        filename: "test-bundle.js",
    },
    devServer: {
        contentBase: "./tests/public",
        port: 8080,
    },
}
```

I have the repo here: https://github.com/marcingolenia/kabanos if You need more information. 

Can You give me some hints on how to move forward with this? Thanks in advance :)

**Comment by davedawkins** [comment]
I started out with console tests, and found that they weren't rich enough for me to check the DOM-related functionality, so I developed a second set of tests that run in the browser.

This means that the Mocha tests are probably out of date. I will check them out later and see if I can bring everything back together under Mocha, perhaps with a headless browser.

I need my tests to run in GH, so this is worth doing.

**Comment by alfonsogarciacaro** [comment]
@davedawkins Lit is only required for the methods that renders HTML and puts it in a disposable container in the DOM. Maybe we could just extract those methods and have a Fable.Expect library that it's independent of Lit: https://github.com/fable-compiler/Fable.Lit/blob/e5f24e282160dbd8a7fd69e79b58592192c76aaf/src/Lit.Test/Expect.Dom.fs#L97-L115

Note that the Expect.XXX modules are only utilities for assertion and dealing with the DOM or Elmish. The headless part is done by [Web Test Runner](https://modern-web.dev/docs/test-runner/overview/). This file contains bindings for working with WTR (like declaring the tests or doing snapshots): https://github.com/fable-compiler/Fable.Lit/blob/e5f24e282160dbd8a7fd69e79b58592192c76aaf/src/Lit.Test/WebTestRunner.fs

**Comment by alfonsogarciacaro** [comment]
> Exactly what I was thinking.

Here you go :) https://github.com/fable-compiler/Fable.Expect

**Comment by alfonsogarciacaro** [comment]
The idea is to use accessible queries so by writing the tests you also make sure your UI is accessible: https://github.com/fable-compiler/Fable.Expect#accessible-queries

Accessible queries will throw if they don't find anything although not an Assertion error. If you want to turn them into assertions you can use `Expect.success`.

```fsharp
container.El
|> Expect.success "new todo found" (fun el -> el.getByText(newTodo))
```

You can use `Expect.successAnd` to chain assertions.

```fsharp
container.El
|> Expect.success "value display" (fun el -> el.getByText("value))
|> Expect.innerText "Value: Foo"
```

If you just want to make sure your UI doesn't change unexpectedly you can use a snapshot test:

```fsharp
do! container.El |> Expect.matchHtmlSnapshot "new-todo"
```

**Comment by davedawkins** [comment]
Yes, and that's what app developers should be doing for sure. This is probably what @marcingolenia needs.

As a framework developer, I need to make sure that the DSL and bindings have the right side effects in the DOM. The accessible queries will help up to a certain level of detail, but not quite enough. (For example, ordering of elements etc).

In any case, this logged issue, and your assistance,  is aimed at the "end-app-developer", so we're in good shape in that regard.

**Comment by alfonsogarciacaro** [comment]
Well, at the end, `container.El` is just and HTMLElement (the accessible query methods are implemented as extensions) so you can still use `.querySelector` or any other native method to inspect the DOM :)

**Comment by davedawkins** [comment]
I've reproduced it with your repo, so looking into it now

**Comment by davedawkins** [comment]
The line that's failing is where you mount the app in App.fs:
```fsharp
app() |> Program.mountElement "sutil-app"
```

In the test environment, the node with ID "sutil-app" doesn't exist and so we get the error.
If you comment that out, your test works.

Going forward, you might want to consider a couple of ways to arrange your project, so that running the tests doesn't try to "run" (mount) the main application.

1. Test for existence of "sutil-app" - use this as a check to decide if you're running as a test or not 

2. Separate your app components into a separate project, and run the tests against that. You can then have a minimal "App" project that instantiates and mounts the main App() class.

I hope this helps!
